specs/stein/approved/amd-sev-libvirt-support.rst

```diff
@@ -0,0 +1,367 @@
+      ...
+    </domain>
+
+For reference, `the AMDSEV GitHub repository
+<https://github.com/AMDESE/AMDSEV/>`_ provides `a complete example of
+SEV-enabled domain XML
+<https://github.com/AMDESE/AMDSEV/blob/master/xmls/sample.xml>`_.
+
+This additional functionality could be implemented within
+``nova.virt.libvirt.LibvirtDriver._set_features()``, e.g. using a new
+``nova.virt.libvirt.LibvirtConfigGuestFeatureSEV`` class.
+
+This work could also mean that images with the property
+``trait:HW_CPU_x86_SEV=required`` would similarly affect the process
+of launching instances.
+
+Immediate follow-up work
+------------------------
+
+It is suggested that follow-up specs be submitted detailing
+the following subsequent work:
+
+- Support a built-in trait which would be automatically detected per
+  compute host by nova's libvirt driver.  This would most likely be
+  called ``HW_CPU_x86_SEV`` or similar, as an extension of `the
+  existing CPU traits mapping
+  <https://github.com/openstack/nova/blob/c5a7002bd571379818c0108296041d12bc171728/nova/virt/libvirt/utils.py#L47>`_.
+
+  Additional logic would be required to check that the various layers
+  of the hardware and software hypervisor stack are *all* SEV-capable:
+
+  - The presence of the following XML in the response from a libvirt
+    `virConnectGetDomainCapabilities()
+    <https://libvirt.org/html/libvirt-libvirt-domain.html#virConnectGetDomainCapabilities>`_
+    API call `indicates that both QEMU and the AMD Secure Processor
+    (AMD-SP) support SEV functionality
+    <https://libvirt.org/git/?p=libvirt.git;a=commit;h=6688393c6b222b5d7cba238f21d55134611ede9c>`_::
+
+        <domainCapabilities>
+          ...
+          <features>
+            ...
+            <sev supported='yes'/>
+              ...
+            </sev>
+          </features>
+        </domainCapabilities>
+
+  - ``/sys/module/kvm_amd/parameters/sev`` should have the value one
+    to indicate that the kernel has SEV capabilities enabled.
+
+- Since `migration of instances between unencrypted and SEV-encrypted
+  states is not supported in either direction
+  <https://github.com/qemu/qemu/commit/8fa4466d77b44f4f58f3836601f31ca5e401485d>`_,
+  prevent nova from resizing an instance from a non-SEV-enabled flavor
+  to an SEV-enabled one, or vice-versa.  Alternatively, nova could
+  catch the error raised by QEMU, which would be propagated via
+  libvirt, and handle it appropriately.
+
+  Similarly, attempts to suspend / resume a domain should either be
+  prevented, or the error caught and handled.
+
+Alternatives
+------------
+
+#. The idea of initially using a custom trait could be skipped in
+   favour of immediately implementing the built-in trait
+   (``HW_CPU_x86_SEV`` or similar).  This would increase the amount of
+   effort required in order to reach a working prototype, but
+   conversely has the advantage of not requiring the merging of
+   unfinished proof-of-concept code into nova's ``master`` branch.
```

Comment:
what do you mean by that? upstream will likely not accept "unfinished proof of concept code". the implementation we're proposing under the spec should be complete with unit test coverage

Reply:
OK, "unfinished" is too alarmist a word - changing that.  It's nothing to do with test coverage; it's about whether it's OK for a reference to a custom trait to be hardcoded in `master`, even for a short time.  Only upstream can answer that.